t_frang/test_http_resp_code_block.py

```diff
@@ -0,0 +1,393 @@
+
+        Why is there no 403 response when the limit is reached?
+        
+        """
+        requests = "GET /uri1 HTTP/1.1\r\n" \
+                   "Host: localhost\r\n" \
+                   "\r\n" * 10
+        requests2 = "GET /uri2 HTTP/1.1\r\n" \
+                    "Host: localhost\r\n" \
+                    "\r\n" * 10
+        nginx = self.get_server('nginx')
+        nginx.start()
+        self.start_tempesta()
+
+        deproxy_cl = self.get_client('deproxy3')
+        deproxy_cl.start()
+
+        deproxy_cl2 = self.get_client('deproxy4')
+        deproxy_cl2.start()
+
+        self.deproxy_manager.start()
+        self.assertTrue(nginx.wait_for_connections(timeout=1))
+
+        deproxy_cl.make_requests(requests)
+        deproxy_cl2.make_requests(requests2)
+
+        deproxy_cl.wait_for_response(timeout=4)
+        deproxy_cl2.wait_for_response(timeout=6)
+
+        self.assertEqual(5, len(deproxy_cl.responses))
+        self.assertEqual(5, len(deproxy_cl2.responses))
+
+        self.assertFalse(deproxy_cl.connection_is_closed())
+        self.assertFalse(deproxy_cl2.connection_is_closed())
+
+
+
+    def test_one_client(self):
+        """
+        One client send irregular chain of 404, 405 and 200 requests with 5 rps. 
+        10 requests: [ '200', '404', '404', '404', '404',
+                       '200', '405', '405', '200', '200']
+        """
+        requests0 = "GET /uri2 HTTP/1.1\r\n" \
+                   "Host: localhost\r\n" \
+                   "\r\n" 
+        requests1 = "GET /uri1 HTTP/1.1\r\n" \
+                   "Host: localhost\r\n" \
+                   "\r\n" * 4
+        requests2 = "GET /uri3 HTTP/1.1\r\n" \
+                   "Host: localhost\r\n" \
+                   "\r\n" * 2
+
+        requests = (requests0)+requests1+requests0+requests2+(requests0*2)
+
+
+        nginx = self.get_server('nginx')
+        nginx.start()
+        self.start_tempesta()
+        deproxy_cl = self.get_client('deproxy2')
+        deproxy_cl.start()
+
+        self.deproxy_manager.start()
+        self.assertTrue(nginx.wait_for_connections(timeout=1))
+
+        deproxy_cl.make_requests(requests)        
+        deproxy_cl.wait_for_response(timeout=4)
+
+        self.assertEqual('403', deproxy_cl.responses[-1].status,
+                         "Unexpected response status code")
+
+        self.assertEqual(8, len(deproxy_cl.responses))        
+        self.assertTrue(deproxy_cl.connection_is_closed())
+
+
+
+    def test_two_clients(self):
+        """Two clients. One client sends 12 requests by 6 per second during
+    2 seconds. Of these, 6 requests by 3 per second give 404 responses.
+    Should be get 11 responses (5 with code 200, 5 with code 404 and
+    1 with code 403).
+    The second client sends 20 requests by 5 per second during 4 seconds.
+    Of these, 10 requests by 2.5 per second give 404 responses. All requests
+    should be get responses.
```

Comment:
First of all, I appreciate the interesting scenarios implemented in the test!

However, it seems there are several issues with the test:
1. the first client sends requests raising 6 error responses, i.e. on the 5th response Tempesta shall block the client and it *should* not receive a response for 6th request with `ip_block on` option.
2. IIRC the deproxy API correctly, `make_requests()` just sends all the pipelined requests in one shot, immediately, and  `wait_for_response()` is just a timeout for which we can wait for a response. I.e. the test is broken.

I reviewed only this one test case. Please rework and fix all the cases in the file.

Reply:
well, now the limit works like this, maybe it's a bug, but it seemed to me that it looks good:
http_resp_code_block 404 405 5 2; <- for more then 5 error request per 2 sec will be blocked
first client case set 6rps:
| Second | Num | Request | Response code | Result |
| :---- | :------ | :------ |  :------- | :------ |
|1 sec | 1 | /uri1 | 404 | error request num 1 per 2 second|
|1 sec | 2 | /uri2 | 200 | ok |
|1 sec | 3 | /uri1 | 404 | error request num 2 per 2 second |
|1 sec  | 4 | /uri2 | 200 | ok |
|1 sec | 5 | /uri1 | 404 | error request num 3 per 2 second |
|1 sec | 6 | /uri2 | 200 | ok |
|2 sec | 7 | /uri1 | 404 | error request num 4 per 2 second |
|2 sec | 8 | /uri2 | 200 | ok |
|2 sec | 9 | /uri1 | 404 | error request num 5 per 2 second |
|2 sec | 10 | /uri2 | 200 | ok |
|2 sec | 11 | /uri1 | 404 | error request num 6  per 2 second - get limit - get 405 code |
|2 sec | 12 | /uri2 | 200 | no resp - client block|

Second client case set 5rps: 

| Second | Num | Request | Response code | Result |
| :---- | :------ | :------ |  :------- | :------ |
|1 sec | 1 | /uri1 | 404 | error request num 1  per 2 second|
|1 sec | 2 | /uri2 | 200 | ok |
|1 sec | 3 | /uri1 | 404 | error request num 2  per 2 second|
|1 sec  | 4 | /uri2 | 200 | ok |
|1 sec | 5 | /uri1 | 404 | error request num 3  per 2 second|
|2 sec | 6 | /uri2 | 200 | ok |
|2 sec | 7 | /uri1 | 404 | error request num 4  per 2 second|
|2 sec | 8 | /uri2 | 200 | ok |
|2 sec | 9 | /uri1 | 404 | error request num 5  per 2 second|
|2 sec | 10 | /uri2 | 200 | ok |
|3 sec | 11 | /uri1 | 404 | error request num 1  per 2 second|
|3 sec | 12 | /uri2 | 200 | ok |
|3 sec | 13 | /uri1 | 404 | error request num 2  per 2 second|
|3 sec | 14 | /uri2 | 200 | ok |
|3 sec | 15 | /uri1 | 404 | error request num 3  per 2 second|
|4 sec | 16 | /uri2 | 200 | ok |
|4 sec | 17 | /uri1 | 404 | error request num 4  per 2 second|
|4 sec | 18 | /uri2 | 200 | ok |
|4 sec | 19 | /uri1 | 404 | error request num 5  per 2 second|
|4 sec | 20 | /uri2 | 200 | ok |

Reply:
I'm still confused: aren't all the pipelined requests just fired immediately? In my understanding of deproxy all the requests in both the tables should have `1 sec` in the leftmost column, so the first assertion `self.assertTrue(deproxy_cl.connection_is_closed())` looks correct, but I don't understand why the second assertion passes... Is there some bug?

Reply:
> I'm still confused: aren't all the pipelined requests just fired immediately?
This test sets rps 5 or 6 for deproxy client. Therefore, client can receive response before sending new request.
I separated test logic for 1 client and multiple. I think it works correctly.